src/Agent.Worker/Build/BuildServer.cs

```diff
@@ -9,28 +9,63 @@
 
 namespace Microsoft.VisualStudio.Services.Agent.Worker.Build
 {
-    public class BuildServer
+    [ServiceLocator(Default = typeof(BuildServer))]
+    public interface IBuildServer : IAgentService
     {
-        private readonly Build2.BuildHttpClient _buildHttpClient;
-        private Guid _projectId;
+        Task ConnectAsync(VssConnection connection);
+        Task<Build2.BuildArtifact> AssociateArtifact(Guid projectId, int buildId, string name, string type, string data, Dictionary<string, string> propertiesDictionary, CancellationToken cancellationToken);
+        Task<Build2.Build> UpdateBuildNumber(Guid projectId, int buildId, string buildNumber, CancellationToken cancellationToken);
+        Task<IEnumerable<string>> AddBuildTag(Guid projectId, int buildId, string buildTag, CancellationToken cancellationToken);
+    }
+
+    public sealed class BuildServer : AgentService, IBuildServer
+    {
+        private bool _hasConnection;
+        private VssConnection _connection;
+        private Build2.BuildHttpClient _buildHttpClient;
 
-        public BuildServer(VssConnection connection, Guid projectId)
+        public async Task ConnectAsync(VssConnection connection)
         {
-            ArgUtil.NotNull(connection, nameof(connection));
-            ArgUtil.NotEmpty(projectId, nameof(projectId));
+            _connection = connection;
+            int attemptCount = 5;
+            while (!_connection.HasAuthenticated && attemptCount-- > 0)
+            {
+                try
+                {
+                    await _connection.ConnectAsync();
+                    break;
+                }
+                catch (Exception ex) when (attemptCount > 0)
+                {
+                    Trace.Info($"Catch exception during connect. {attemptCount} attempt left.");
+                    Trace.Error(ex);
+                }
+
+                await Task.Delay(100);
+            }
+
+            _buildHttpClient = _connection.GetClient<Build2.BuildHttpClient>();
+            _hasConnection = true;
```

Comment:
Can we hit this without having a connection? If _connection.HasAuthenticated stays false and we run out of attempts we would hit this line and not have an authenticated connection. We may need to check HasAuthenticated again and set _hasConnection based on that? Unless I am misunderstanding the use of _hasConnection :)

EDIT: It's somewhat subtle that ConnectAsync will throw on failure. Can we add a few comments? Or maybe change the response type to boolean? If we are trying to connect it doesn't seem exceptional for the connection to fail.